File: src/main/java/org/kite9/diagram/visualization/planarization/mapping/ContainerLayoutEdge.java
package org.kite9.diagram.visualization.planarization.mapping;

import org.kite9.diagram.common.elements.AbstractPlanarizationEdge;
import org.kite9.diagram.common.elements.PlanarizationEdge;
import org.kite9.diagram.common.elements.Vertex;
import org.kite9.diagram.position.Direction;
import org.kite9.diagram.primitives.Connected;
import org.kite9.diagram.primitives.DiagramElement;

/**
 * This edge is used to enforce a particular layout within a container.
 * 
 * It is used between the sibling contents of the container.
 * 
 * @author robmoffat
 *
 */
public class ContainerLayoutEdge extends AbstractPlanarizationEdge {
	
	GeneratedLayoutElement underlying;
	
	public ContainerLayoutEdge(Vertex from, Vertex to, Direction d, Connected fromElement, Connected toElement) {
		super(from, to, null, null, null, null, null);
		this.drawDirection = d;
		this.underlying =  new GeneratedLayoutElement((Connected) from.getOriginalUnderlying(), (Connected) to.getOriginalUnderlying(), d);
	}
	
	private ContainerLayoutEdge(Vertex from, Vertex toIntroduce, Direction drawDirection, boolean reversed, GeneratedLayoutElement underlying) {
		super(from, toIntroduce, null, null, null, null, null);
		this.reversed =true;
		this.underlying = underlying;
		this.drawDirection = drawDirection;
	}

	public DiagramElement getOriginalUnderlying() {
		return underlying;
	}

	@Override
	public int getCrossCost() {
		return 0;	// no cost for traversing between items in the layout
	}

	@Override
	public RemovalType removeBeforeOrthogonalization() {
		// can be removed if there is another edge to do the same job
		return RemovalType.TRY;	
	}

	public boolean isLayoutEnforcing() {
		return true;
	}

	public void setLayoutEnforcing(boolean le) {
		throw new UnsupportedOperationException("Layout edges are always layout enforcing");
	}

	@Override
	public PlanarizationEdge[] split(Vertex toIntroduce) {
		PlanarizationEdge[] out = new PlanarizationEdge[2];
		out[0] = new ContainerLayoutEdge(getFrom(), toIntroduce, getDrawDirection(), isReversed(), underlying);
		out[1] = new ContainerLayoutEdge(toIntroduce, getTo(),  getDrawDirection(), isReversed(), underlying);

		return out;
	}

	@Override
	public boolean isReversed() {
		return false;
	}

	@Override
	public int getLengthCost() {
		return 0;
	}
}

File: visualization/kite9-visualization-common/src/main/kotlin/org/kite9/diagram/common/elements/mapping/ContainerLayoutEdge.kt
package org.kite9.diagram.common.elements.mapping

import org.kite9.diagram.common.elements.edge.AbstractPlanarizationEdge
import org.kite9.diagram.common.elements.edge.BiDirectionalPlanarizationEdge
import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.edge.PlanarizationEdge.RemovalType
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.common.objects.Pair
import org.kite9.diagram.model.Connected
import org.kite9.diagram.model.ConnectedRectangular
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.position.Direction

/**
 * This edge is used to enforce a particular layout within a container.
 *
 * It is used between the sibling contents of the container.
 *
 * @author robmoffat
 */
class ContainerLayoutEdge(
    f: Vertex,
    t: Vertex,
    dd: Direction,
    val underlying: GeneratedLayoutBiDirectional,
    override var fromUnderlying: DiagramElement?,
    override var toUnderlying: DiagramElement?) :
    AbstractPlanarizationEdge(f, t, dd),
    BiDirectionalPlanarizationEdge {

    constructor(from: Vertex, to: Vertex, d: Direction, fromElement: Connected, toElement: Connected) : this(
        from,
        to,
        d,
        GeneratedLayoutBiDirectional(fromElement, toElement, d),
        fromElement,
        toElement
    ) {
    }

    override fun getOriginalUnderlying(): DiagramElement {
        return underlying
    }

    override fun getCrossCost(): Int {
        return 0 // no cost for traversing between items in the layout
    }

    override fun removeBeforeOrthogonalization(): RemovalType {
        // can be removed if there is another edge to do the same job
        return RemovalType.TRY
    }

    override fun isLayoutEnforcing(): Boolean {
        return true
    }

    override fun setLayoutEnforcing(le: Boolean) {
        throw UnsupportedOperationException("Layout edges are always layout enforcing")
    }

    override fun split(toIntroduce: Vertex): Pair<PlanarizationEdge> {
        val out = Pair<PlanarizationEdge>(
            ContainerLayoutEdge(
                getFrom(),
                toIntroduce,
                getDrawDirection(),
                underlying,
                fromUnderlying,
                null),
            ContainerLayoutEdge(
                toIntroduce,
                getTo(),
                getDrawDirection(),
                underlying,
                null,
                toUnderlying));
        return out
    }

    override fun isPartOf(de: DiagramElement?): Boolean {
        return getOriginalUnderlying() === de
    }

    override fun getDiagramElements(): MutableMap<DiagramElement, Direction?> {
        return mutableMapOf(kotlin.Pair(getOriginalUnderlying(), null))
    }

    override fun getDrawDirection() : Direction {
        return super.getDrawDirection()!!
    }

    init {
        setStraight(true)
    }
}